helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -403,16 +403,55 @@ public RoutingTableSnapshot getRoutingTableSnapshot(PropertyType propertyType, S
     return snapshots;
   }
 
+
   /**
    * Add RoutingTableChangeListener with user defined context
    * @param routingTableChangeListener
    * @param context user defined context
    */
   public void addRoutingTableChangeListener(
       final RoutingTableChangeListener routingTableChangeListener, Object context) {
+    addRoutingTableChangeListener(routingTableChangeListener, context, false);
+  }
+
+  /**
+   * Add RoutingTableChangeListener with user defined context
+   * @param routingTableChangeListener
+   * @param context user defined context
+   * @param doInit whether to trigger the initial callback during adding listener
+   */
+  public void addRoutingTableChangeListener(
```

Comment:
This is a high-level question. Have we considered doing init as the default behavior? Here are the reasons I can think off the top of my head:

1. It doesn't hurt to init (what are the bad side effects of reading in the beginning? If the user wishes to add a listener, that means the user is ready to read and be notified). In other words, I can't think of any negative consequences of this behavior change or any potential backward-compatibility issues that might break things.

2. This greatly reduces verbosity in the code. Simple is better.

Reply:
@dasahcc had some concerns that if users are not expecting an empty snapshot, and the initial callback provides them such, they may not have a way to handle it yet.

Reply:
I'm not quite seeing that - applications should program in a way that expects some data in the snapshot, right? Their logic shouldn't try to "time it" in their callback logic since that's never predictable.

Reply:
let's quickly sync offline.